alembic/versions/a3a7212e29b8_source_add_phonebook_uuid.py

```diff
@@ -0,0 +1,139 @@
+    sa.column('tenant_uuid'),
+)
+
+
+def upgrade():
+    op.add_column(
+        'dird_source',
+        sa.Column(
+            'phonebook_uuid',
+            postgresql.UUID(as_uuid=False),
+            sa.ForeignKey('dird_phonebook.uuid', ondelete='CASCADE'),
+            nullable=True,
+        ),
+    )
+    op.create_unique_constraint(
+        'dird_phonebook_tenant_uuid_idx',
+        'dird_phonebook',
+        ['uuid', 'tenant_uuid'],
+    )
+    op.create_foreign_key(
+        'dird_source_phonebook_fkey',
+        'dird_source',
+        'dird_phonebook',
+        ['phonebook_uuid', 'tenant_uuid'],
+        ['uuid', 'tenant_uuid'],
+        ondelete='CASCADE',
+    )
+    phonebook_sources_query = sa.select(
+        [
+            source_table.c.uuid,
+            source_table.c.extra_fields,
+        ]
+    ).where(source_table.c.backend == 'phonebook')
+    phonebooks_query = sa.select(
+        [
+            phonebook_table.c.uuid,
+            phonebook_table.c.name,
+            phonebook_table.c.id,
+        ]
+    )
+    phonebooks = list(op.get_bind().execute(phonebooks_query))
+    phonebook_by_name = {p.name: p for p in phonebooks}
+    phonebook_by_id = {p.id: p for p in phonebooks}
+    phonebook_uuid_map = {}
+    for row in op.get_bind().execute(phonebook_sources_query):
+        if 'phonebook_uuid' in row.extra_fields:
+            phonebook_uuid_map[row.uuid] = row.extra_fields['phonebook_uuid']
+        elif 'phonebook_id' in row.extra_fields:
+            phonebook_id = row.extra_fields['phonebook_id']
+            try:
+                phonebook = phonebook_by_id[phonebook_id]
+                phonebook_uuid_map[row.uuid] = phonebook.uuid
+            except KeyError:
+                print(
+                    'phonebook(id={phonebook_id}) not found for source {row.uuid}. Future migration might delete it.'
+                )
+                continue
+        else:
+            try:
+                phonebook = phonebook_by_name[row.name]
+                phonebook_uuid_map[row.uuid] = phonebook.uuid
+            except KeyError:
+                print(
+                    f'could not map source {row.uuid} to a phonebook. Future migration might delete it.'
+                )
+                continue
+
+    print('source to phonebook: ', phonebook_uuid_map)
+    if phonebook_uuid_map:
+        phonebook_sources_update = (
+            source_table.update()
+            .where(
+                source_table.c.uuid == sa.bindparam('_source_uuid'),
+            )
+            .values(
+                phonebook_uuid=sa.bindparam('_phonebook_uuid'),
+                extra_fields=(
+                    sa.cast(source_table.c.extra_fields, postgresql.JSONB)
+                    - 'phonebook_uuid'
+                ),
+            )
+        )
+        op.get_bind().execute(
+            phonebook_sources_update,
+            [
+                {'_source_uuid': uuid, '_phonebook_uuid': phonebook_uuid}
```

Comment:
Out of curiosity, why the leading `_`?

Reply:
Honestly not sure, I think I feared namespace clashes after seeing some logs.